pkgs/development/python-modules/scipy/default.nix

```diff
@@ -19,10 +21,15 @@
 , libxcrypt
 }:
 
+assert blas.provider == numpy.blas;
```

Comment:
no asserts, because they can block overriding.

Reply:
I didn't use `blas.provider` when I implemented multiple `blas` overrides in `octave`, because there things are a little bit more complicated - the `octave` expression has to coordinate between many dependencies that depend and should use the same `blas` and `lapack`. In this case, it's much simpler to my understanding, and I don't understand what's wrong with using `numpy.blas` which points to `blas.provider`. In the current state of things you can just:

```nix
scipy-myblas = super.scipy.override {
  numpy = super.numpy.override {
    blas = self.myblas;
  };
};
```

Which seems very nice to me.

Reply:
> In the current state of things you can just ... @doronbehar 

Personally, I find this rather obscure: blas is scipy's direct dependency, but we cannot just override it, we have to override numpy instead. The actual "reason" I'm making `blas` an explicit argument, however, is that from `numpy.blas` (evaluates into `mkl` in case of `blas = prev.blas.override { blasProvider = final.mkl; }`) we cannot infer the correct pkg-config target name. With the blas-switching derivation the pkg-config target is fixed at `cblas`, as far as I can tell.

The reason I put the `assert` is because previously it wasn't possible to have un-synchronized `numpy.blas` and `scipy.blas`, but with the new interface it is and even likely to happen by accident (e.g. if one had a local override like your `scipy-myblas`, then it could still evaluate, but scipy its propagated numpy would be silently using different BLAS implementations)

I see the convenience argument, but global overlays are same cost in LOC (assuming there's a binary cache), are safer, and this PR doesn't break them.

I also haven't looked into why `numpy` exposes `blas.provider` instead of just `blas`.

Further thoughts?

@FRidh I'll switch to `meta.broken`, and later open a PR to migrate cuda packages as well

Reply:
> Personally, I find this rather obscure: blas is scipy's direct dependency, but we cannot just override it, we have to override numpy instead.

You are right - if a scipy dependent that also depends on `blas` will want to override `blas` it will have to:

```nix
scipy-dependent = super.scipy-dependent.override {
  scipy = super.scipy.override {
    numpy = super.numpy.override {
      blas = self.myblas;
    };
  };
};
```

And this indeed escalates. I have a different proposal that doesn't invlove asserts, with a similar thinking to what's happening in `octave`:

```patch
diff --git i/pkgs/development/python-modules/scipy/default.nix w/pkgs/development/python-modules/scipy/default.nix
index 1090e724a7f..efa80fced17 100644
--- i/pkgs/development/python-modules/scipy/default.nix
+++ w/pkgs/development/python-modules/scipy/default.nix
@@ -14,12 +14,17 @@
 , pytest
 , pytest-xdist
 , numpy
+, blas
 , pybind11
 , pooch
 , libxcrypt
 }:
 
-buildPythonPackage rec {
+let
+  numpy' = numpy.override {
+    inherit blas;
+  };
+in buildPythonPackage rec {
   pname = "scipy";
   version = "1.10.1";
   format = "pyproject";
@@ -39,14 +44,14 @@ buildPythonPackage rec {
   nativeBuildInputs = [ cython gfortran meson-python pythran pkg-config wheel ];
 
   buildInputs = [
-    numpy.blas
+    blas
     pybind11
     pooch
   ] ++ lib.optionals (pythonOlder "3.9") [
     libxcrypt
   ];
 
-  propagatedBuildInputs = [ numpy ];
+  propagatedBuildInputs = [ numpy' ];
 
   nativeCheckInputs = [ nose pytest pytest-xdist ];
 
@@ -79,7 +84,8 @@ buildPythonPackage rec {
   requiredSystemFeatures = [ "big-parallel" ]; # the tests need lots of CPU time
 
   passthru = {
-    blas = numpy.blas;
+    inherit blas;
+    numpy = numpy';
   };
 
   setupPyBuildFlags = [ "--fcompiler='gnu95'" ];
```

Reply:
> You are right - if a scipy dependent that also depends on `blas` will want to override `blas` it will have to:
> 
> ```nix
> scipy-dependent = super.scipy-dependent.override {
>   scipy = super.scipy.override {
>     numpy = super.numpy.override {
>       blas = self.myblas;
>     };
>   };
> };
> ```
> 
> And this indeed escalates

You can override blas for the python package set only.

Reply:
> You can override blas for the python package set only.

How?

Reply:
If I recall correctly setting `blas` on the Python package set *should* make it being by all `callPackage` inside of the package set.

Reply:
> If I recall correctly setting `blas` on the Python package set _should_ make it being used by all `callPackage` inside of the package set.

You mean something like the following?

```
pythonWithMyBlas = python3.override {
  packageOverrides = self: super: {
    blas = super.pkgs.blas.override {
      # ... 
    };
  };
};
```

Reply:
Exactly!

And don't forget the `self = pythonWithMyBlas;`.

Reply:
Only I think `blas` and `lapack` come from the outer package set (`nixpkgs#blas`, not `nixpkgs#python3Packages.blas`). But the idea is the same, only that it's even coarser granularity: you overlay the entire nixpkgs to achieve mutual compatibility. We can also add `meta.broken = blas != numpy.blas`. I still don't see any reason we expose `blas.provider` instead of `blas` in `numpy.passthru`, so I think we shouldn't be doing that

RE: `python3.override`

```
  pythonPackagesExtensions = prev.pythonPackagesExtensions ++ [
    (python-final: python-prev: { ... })
  ];
```

Reply:
> Only I think blas and lapack come from the outer package set (nixpkgs#blas, not nixpkgs#python3Packages.blas).

CallPackage should pick it up if I am correct

> I still don't see any reason we expose blas.provider instead of blas in numpy.passthru, so I think we shouldn't be doing that

I don't know anymore why that was. Maybe that should be changed throughout?

> Only I think `blas` and `lapack` come from the outer package set (`nixpkgs#blas`, not `nixpkgs#python3Packages.blas`). But the idea is the same, only that it's even coarser granularity: you overlay the entire nixpkgs to achieve mutual compatibility. We can also add `meta.broken = blas != numpy.blas`. I still don't see any reason we expose `blas.provider` instead of `blas` in `numpy.passthru`, so I think we shouldn't be doing that
> 
> RE: `python3.override`
> 
> ```
>   pythonPackagesExtensions = prev.pythonPackagesExtensions ++ [
>     (python-final: python-prev: { ... })
>   ];
> ```

That would have an effect on the entire nixpkgs then (all interpreters + every package using any of these).

Reply:
Ah, so we could just _add_ the `blas` attribute to the python package set, is that what you're saying? And in either case it would be picked up by `callPackage`.

Reply:
Exactly!